dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java

```diff
@@ -0,0 +1,277 @@
+    {
+        Objects.requireNonNull( trackedEntity, Constants.TRACKED_ENTITY_CANT_BE_NULL );
+
+        Map<String, TrackedEntityAttributeValue> valueMap = new HashMap<>();
+        if ( tei != null )
+        {
+            valueMap = tei.getTrackedEntityAttributeValues()
+                .stream()
+                .collect( Collectors.toMap( v -> v.getAttribute().getUid(), v -> v ) );
+        }
+
+        for ( Attribute attribute : trackedEntity.getAttributes() )
+        {
+            TrackedEntityAttribute tea = reporter.getValidationContext()
+                .getTrackedEntityAttribute( attribute.getAttribute() );
+
+            if ( tea == null )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1006 )
+                    .addArg( attribute.getAttribute() ) );
+                continue;
+            }
+
+//            if ( StringUtils.isEmpty( attribute.getValue() ) )
+            if ( attribute.getValue() == null )
+            {
+                //continue; ??? Just continue on empty and null?
+                // TODO: Is this really correct? This check was not here originally
+                //  Enrollment attr check fails on null so why not here too?
+                reporter.addError( newReport( TrackerErrorCode.E1076 )
+                    .addArg( attribute ) );
+                continue;
+            }
+
+            // TODO: Should we really validate existing data? this sounds like a mix of con
+//            TrackedEntityAttributeValue trackedEntityAttributeValue = valueMap.get( tea.getUid() );
+//            if ( trackedEntityAttributeValue == null )
+//            {
+            TrackedEntityAttributeValue trackedEntityAttributeValue = new TrackedEntityAttributeValue();
+            trackedEntityAttributeValue.setEntityInstance( tei );
+            trackedEntityAttributeValue.setValue( attribute.getValue() );
+            trackedEntityAttributeValue.setAttribute( tea );
+//            }
+
+            validateAttributeValue( reporter, trackedEntityAttributeValue );
+            validateTextPattern( reporter, attribute, tea, valueMap.get( tea.getUid() ) );
+            validateAttrValueType( reporter, attribute, tea );
+
+            // TODO: This is one "THE" potential performance killer...
+            validateAttributeUniqueness( reporter, attribute.getValue(), tea, tei, orgUnit );
+
+            validateFileNotAlreadyAssigned( reporter, attribute, valueMap );
+        }
+    }
+
+    public void validateAttributeValue( ValidationErrorReporter reporter, TrackedEntityAttributeValue teav )
+    {
+        Objects.requireNonNull( teav, TRACKED_ENTITY_ATTRIBUTE_VALUE_CANT_BE_NULL );
+        Objects.requireNonNull( teav.getValue(), TRACKED_ENTITY_ATTRIBUTE_VALUE_CANT_BE_NULL );
+
+        // TODO: Should this be tested here, cant provoke? Is this not validation in metadata? Probably not according to Stian.
```

Comment:
`TrackedEntityAttribute.valueType` has `not-null=true` so I don't think we need to double verify anything here